turftopic/encoders/e5.py

```diff
@@ -30,11 +30,64 @@ class E5Encoder(SentenceTransformer, ExternalEncoder):
     """
 
     def __init__(
-        self, model_name: str = "intfloat/multilingual-e5-small", instruction: Optional[str] = None, **kwargs
+        self, model_name: str = "intfloat/multilingual-e5-small", **kwargs
+    ):
+        super().__init__(model_name, **kwargs)
+
+    def encode(self, sentences: list[str], **kwargs) -> np.ndarray:
+        """
+        Parameters
+        ----------
+        sentences: list[str]
+            Input text
+
+        Notes
+        -----
+        See docs for `SentenceTransformer.encode` for available **kwargs  
+        https://www.sbert.net/docs/package_reference/SentenceTransformer.html#sentence_transformers.SentenceTransformer.encode
+        """
+        sentences = [
+            f"query: {sentence}" for sentence in sentences
+        ]
+        return super().encode(sentences, **kwargs)
+
+
+class E5InstructionalEncoder(SentenceTransformer):
```

Comment:
@jankounchained I'm not sure I understand the rationale behind separating this into two classes. Can't we just roll with one class, where the instruction is optional? As far as I understand it would have the same behaviour, and the only reason I see behind splitting these two is so that there can be a different default. (which I already commented on previously)

Also my guess is that you thought the interface would make sense, because users care about what the encoder does for them, not how it is implemented in the background. But then poor people will think that E5Instruction is in a different class because it can use the Mistral architecture encoders(which it can't).

My suggestion here would be to have one class, for E5 based on sentence transformers (BERT architecture) with an optional instruction `E5Encoder` and one on Mistral `E5MistralEncoder` with a default instruction.
Then we can include documentation explaining what the difference is, and which models need what.

What do you think @jankounchained ?

Reply:
No, I'm not only splitting them because of the default models.
From what I could see in the model cards, the prefixes are different between instruct and non-instruct E5 models.

See e5-small 
https://huggingface.co/intfloat/multilingual-e5-small
```
# Each input text should start with "query: " or "passage: ", even for non-English texts.
# For tasks other than retrieval, you can simply use the "query: " prefix.

input_texts = ['query: how much protein should a female eat',
...
```

While the instruct model does this 
https://huggingface.co/intfloat/multilingual-e5-large-instruct
```
def get_detailed_instruct(task_description: str, query: str) -> str:
    return f'Instruct: {task_description}\nQuery: {query}'

task = 'Given a web search query, retrieve relevant passages that answer the query'
queries = [
    get_detailed_instruct(task, 'how much protein should a female eat'),
...
```

I agree it's unwieldy though.

Reply:
We could merge the two classes & add an argument to let the user define the prefixing style (with a couple of pre-defined options). Leave it open.